### PR TITLE
Installing Python Package developed to provide user pathways to access curated set of chapters

### PR DESCRIPTION
**Pull request by arya1302**
### Summary

Hi all, I am Arya, a GSoC student working with The Turing Way team for the year 2023. 
You can find all updates from my work on this discussion: https://github.com/the-turing-way/the-turing-way/discussions/3081.

This pull request is associated with #2246 

### List of changes proposed in this PR (pull-request)

* ~~I have tested and improved the Python package developed by The Turing Way team on this repo: https://github.com/alan-turing-institute/bio-Turing-Way/~~
* ~~This pull request installs the package~~
* ~~The biggest change this will pose is to build the book with ```python pathways.main pathways /path/to/book command```.~~
* Add [pathways v0.1.0](https://github.com/the-turing-way/pathways/tree/v0.1.0) to requirements
* Add Makefile with commands for book build and pathways build
* Change Netlify build to pathways build
* Add test for pathways build to CI

### What should a reviewer concentrate their feedback on?

- [ ]  Any errors that occur during the book-building process?.
- [ ] Any issues present in the website.

### Acknowledging contributors

- [ ] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/main/README.md#contributors) in the README file.
- [ ] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/main/README.md#contributors) in the README file: 
  - Contributors to the pathways Python package for infrastructure, code

### CODE REVIEW

**Comment by malvikasharan** [comment]
@arya1302 can you also help update the local deployment instructions from here: https://github.com/alan-turing-institute/bio-Turing-Way/blob/main/pathways/documentation.md to https://github.com/the-turing-way/the-turing-way/blob/main/book/README.md

@JimMadge @da5nsy @sgibson91 @bsipocz -- tagging you for your awareness, as this installation will affect some major changes in the deployment workflow.

File: book/website/profiles.yml
Comment (da5nsy) [inline book/website/profiles.yml:15]: Lorem ipsum?

File: book/website/communication/binder/zero-to-binder.md
Comment (da5nsy) [inline book/website/communication/binder/zero-to-binder.md:23]: Is this intended?

File: book/website/communication/citable/citable-cite.md
Comment (da5nsy) [inline book/website/communication/citable/citable-cite.md:61]: Is this intended?

File: book/website/communication/citable/citable-cite.md
Comment (da5nsy) [inline book/website/communication/citable/citable-cite.md:71]: Is this intended?

File: netlify.toml
Comment (da5nsy) [inline netlify.toml:4]: Why does this change?

**Comment by da5nsy** [comment]
This is very cool!

**Comment by da5nsy** [comment]
To aid my understanding of this: what is the benefit of this way of doing it (with an external repo and JavaScript) compared to simply using static links and headers on the index page?

**Comment by malvikasharan** [comment]
@da5nsy I am resolving your comments as this issue is not ready for review -- @likeajumprope will be reviewing this opening for review by others. The files here are for testing -- and in the next day will be cleaned up for merging.

File: netlify.toml
Comment (malvikasharan) [reply netlify.toml:4]: This is because the feature applied here for pathway creation is through a package we developed last year. Arya has been working over the last few months to improve its usability as deployed here: https://deploy-preview-3277--the-turing-way.netlify.app/#different-pathways.

**Comment by malvikasharan** [comment]
> To aid my understanding of this: what is the benefit of this way of doing it (with an external repo and JavaScript) compared to simply using static links and headers on the index page?
I suggest checking these proposals for:
- why we wanted to implement the persona and pathway feature on the book, and the decisions taken by the development team last year: https://github.com/the-turing-way/the-turing-way/blob/main/project_management/2021-07-ux-funding-turing.md -- discussion of development decision here: https://github.com/the-turing-way/the-turing-way/issues/2071
- what Arya has been doing to improve the software: https://github.com/the-turing-way/the-turing-way/pull/3292 (proposal is in this PR)

Furthermore, you can test the implementation here: https://deploy-preview-3277--the-turing-way.netlify.app/#different-pathways. This is not currently possible through JupyterBook, and developing this within JupyterBook was not possible due to the maintenance burden on the JupyterBook team. If we can implement it here in The Turing Way, we can discuss the upstream contribution then.

File: book/website/communication/citable/citable-cite.md
Comment (malvikasharan) [reply book/website/communication/citable/citable-cite.md:71]: This will be updated. Currently in the test mode.

File: book/website/communication/citable/citable-cite.md
Comment (malvikasharan) [reply book/website/communication/citable/citable-cite.md:61]: @arya1302 can you explain why this page has been updated? Or as Danny is asking, if the edit was intended?

File: book/website/profiles.yml
Comment (malvikasharan) [reply book/website/profiles.yml:15]: This will be updated -- in fact, I have a local file that needs to be pushed here.

Comment (malvikasharan) [reply book/website/profiles.yml]: Updated the file.

File: book/website/communication/binder/zero-to-binder.md
Comment (malvikasharan) [reply book/website/communication/binder/zero-to-binder.md:23]: Don't know -- I assume, Arya can explain this.

File: netlify.toml
Comment (da5nsy) [reply netlify.toml:4]: Sorry, I should've been clearer: why `website_copy/_build` rather than `_build`

File: book/website/communication/citable/citable-cite.md
Comment (malvikasharan) [reply book/website/communication/citable/citable-cite.md:71]: I assume not -- I suspect it's an edit slipped into this PR.

File: book/website/communication/citable/citable-cite.md
Comment (malvikasharan) [reply book/website/communication/citable/citable-cite.md:61]: On second though, I think it's an edit that might have slipped unintentionally in this PR.

File: netlify.toml
Comment (malvikasharan) [reply netlify.toml:4]: Good point! I will let @arya1302 and @likeajumprope address that. Thank you for pointing to that.

File: book/website/communication/binder/zero-to-binder.md
Comment (arya1302) [reply book/website/communication/binder/zero-to-binder.md:23]: No intentional edits were made here. There was a version conflict. The main project uses Jupyter Book version 0.13, but the pathway feature [developed here](https://github.com/alan-turing-institute/bio-Turing-Way/tree/main) requires version 0.11.2. While dealing with the conflict, this edit accidentally became a part of it.

File: book/website/communication/citable/citable-cite.md
Comment (arya1302) [reply book/website/communication/citable/citable-cite.md:61]: Yes, this edit slipped into this PR, and it was not intended.

File: netlify.toml
Comment (arya1302) [reply netlify.toml:4]: As part of the pathway feature implementation, the book is built into the 'website_copy' directory (please check lines 59, 60, and 87 of[ https://github.com/alan-turing-institute/bio-Turing-Way/blob/main/pathways/pathways/main.py](https://github.com/alan-turing-institute/bio-Turing-Way/blob/main/pathways/pathways/main.py)). The site contents will only be found in the 'website_copy/_build/html' directory. So this edit was necessary for deploying the site to Netlify.

File: book/website/_config.yml
Comment (sgibson91) [inline book/website/_config.yml:62]:
```suggestion
    suppress_warnings       : ['app.add_node','app.add_directive','app.add_role']
```

See my comment above about splitting these onto their own lines

File: netlify.toml
Comment (sgibson91) [reply netlify.toml:4]: But if the intention is to make this the default way of deploying the book, why not remove the "new_path" stuff in the Python file, and continue to use the current directory?

Comment (sgibson91) [reply netlify.toml:4]: > Arya is testing the best way to deploy the package

_from https://github.com/the-turing-way/the-turing-way/pull/3277#issuecomment-1722568110_

Maybe we are still in the testing phase

**Comment by sgibson91** [comment]
FYI @malvikasharan next time you want a review from the infrastructure team 😉 

Screenshot 2023-09-18 at 13 50 23

File: book/website/communication/binder/zero-to-binder.md
Comment (malvikasharan) [reply book/website/communication/binder/zero-to-binder.md:23]:
```suggestion
```

File: book/website/communication/citable/citable-cite.md
Comment (malvikasharan) [reply book/website/communication/citable/citable-cite.md:61]:
```suggestion
```

File: book/website/communication/citable/citable-cite.md
Comment (malvikasharan) [reply book/website/communication/citable/citable-cite.md:70]:
```suggestion
```

**Comment by malvikasharan** [comment]
I can't figure out why the last pathway for software citation is not deploying as intended (as for other pathways). I am checking if there is any yml file issue. Can anyone online here help me debug this? 😅 -- the classic live demo failure.

File: netlify.toml
Comment (arya1302) [reply netlify.toml:4]: > But if the intention is to make this the default way of deploying the book, why not remove the "new_path" stuff in the Python file, and continue to use the current directory?

This was implemented in 2022 when the Python package was developed. I'm not entirely sure about the specific reasons behind it. If necessary, I can certainly try removing the 'new_path' and use the current directory instead.

**Comment by malvikasharan** [comment]
Thanks @arya1302 for noticing what was wrong -- I would have never figured it out.
So turns out that each profile needs `- reproducible-research/reproducible-research` to provide the top header under which all the curated chapters sit.

We should figure out why that is a requirement -- and how we can change it. No urgency on this as we can definitely come back to this after your reporting stage is over.

**Review by likeajumprope — APPROVED** [review]

File: netlify.toml
Comment (likeajumprope) [reply netlify.toml:4]: maybe we could already make the pathways deployment the "default" way but keep the old way with "old_path"..

Comment (arya1302) [reply netlify.toml:4]: I will try removing the website_copy directory and see if everything works fine.

**Comment by malvikasharan** [comment]
@likeajumprope @arya1302 what is the status of this PR? I am curious to move this forward before the Book Dash. 💟

**Comment by arya1302** [comment]
> @likeajumprope @arya1302 what is the status of this PR? I am curious to move this forward before the Book Dash. 💟

@malvikasharan , the suggested changes have been implemented.

**Comment by malvikasharan** [comment]
Tagging @the-turing-way/infrastructure to ask for their approval on this. Especially, what needs to be updated before/after merging this.

**Comment by sgibson91** [comment]
My main concern around merging this right now is an increase in complexity adding to the maintenance burden of an already under-resourced infrastructure group. It would be really nice to have some onboarding for the team so that if/when things go wrong, we have the right understanding to be able to fix this. Because obviously folk will bring a failure in this feature to the infra team even though we didn't develop it. (I don't mean that sarcastically at all, it is a very understandable assumption that this is infrastructure that we'd be involved in.)

I have just been putting out Netlify fires though, so maybe I'm super sensitive to a new hunk of tech landing in main right now.

File: book/website/requirements.txt
Comment (JimMadge) [inline book/website/requirements.txt:11]: Now that we have an org. It would be good to take that package and move it to a new repository in our org.

Perhaps it is a project the infrastructure WG should adopt and maintain?

File: book/website/profiles.yml
Comment (JimMadge) [inline book/website/profiles.yml:102]:
```suggestion
  description: From research to government, every sector strongly relies upon the
```

File: book/website/profiles.yml
Comment (JimMadge) [inline book/website/profiles.yml:78]:
```suggestion
  description: Research Software Engineers are programmers working in
    an academic context who play crucial roles in projects involving data and
```

Want to be careful not to emphasise science/STEM or the background of an RSE, more what they are doing _now_.

**Review by JimMadge — CHANGES_REQUESTED** [review]
This is looking great in the preview :rocket:.
I think the code/format changes above have been addressed.

A couple of text suggestions.

Slightly in two minds about moving the Python package. I think it is definitely the right thing to do, but I don't want to block this.

File: book/website/requirements.txt
Comment (sgibson91) [reply book/website/requirements.txt:11]: > Perhaps it is a project the infrastructure WG should adopt and maintain?

I think a lot of my concern in my above comment is that it will be assumed anyway that the infra WG are maintaining it - so we need to be onboarded to it, but we are also stretched thin.

**Comment by JimMadge** [comment]
@sgibson91's points are great. It is some additional burden. If it breaks, who will fix it? It could (should?) be the infrastructure team, but we don't have any experience with the code.

Might be worth having a higher bandwidth conversation about these things.

- Who maintains this
- Do we have fallback for when it breaks (say, deploy the book without pathways)

I think I'm keen to support this as the pathways look like a great addition to create digests of the book for readers.

**Comment by sgibson91** [comment]
Yeah, my comments are in no way meant to be taken as opposition to supporting pathways, but there needs to be some more work done to answer maintenance responsibility questions first.

File: book/website/profiles.yml
Comment (malvikasharan) [reply book/website/profiles.yml:78]: Great point! Thank you.

**Comment by malvikasharan** [comment]
Let me check with the developers of this package if one/some of them would like to be named as maintainers of the package. I don't expect the infra team to maintain them. 

I do want this feature to be deployed as it has been quite important for driving some of our conversations with different communities, and it has taken a lot of work by many community members which I would like to support. I do hope we have some folks willing to maintain it outside the Infra WG. 🤞🏽

File: book/website/requirements.txt
Comment (malvikasharan) [reply book/website/requirements.txt:11]: Thanks both. I appreciate the feedback. My read of the situation is if we can find a maintainer outside the infra WG, that would make the decision to merge this feature easier. Is that right?

**Comment by malvikasharan** [comment]
This is a good point. 
> Do we have fallback for when it breaks (say, deploy the book without pathways)

@arya1302 -- have you considered this in your workflow? (I am sure you mentioned this in one of our earlier meetings)

File: book/website/requirements.txt
Comment (JimMadge) [reply book/website/requirements.txt:11]: @malvikasharan my thoughts are, The Turing Way should maintain the code and the infra WG seem like the best people to do that.

I think _we_ should invest in it because it adds value to the book and we will depend on it. Longer term, it would be really great if we build on the work to make a more general pathways plugin for Jupyterbook/Sphinx.

Like @sgibson91 said having some onboarding is probably the most important thing, so that we are best prepared to maintain it. I feel fairly comfortable with the infra WG taking responsibility for it. I think it's certainly a better situation than hoping or expecting someone outside of the project will continue to support it.

**Comment by JimMadge** [comment]
I've tried splitting the Python package into a repo [here](https://github.com/the-turing-way/pathways).

[This tag](https://github.com/the-turing-way/pathways/tree/v0.1.0) should be the same as the head of the main branch on the other repo.

File: book/website/requirements.txt
Comment (JimMadge) [reply book/website/requirements.txt]: Updated this in ab89137 to point to https://github.com/the-turing-way/pathways v0.1.0.

Looks fine to me in the preview.

**Review by sgibson91 — APPROVED** [review]

**Review by JimMadge — APPROVED** [review]

**Pull request merged**
